File: src/midi.zig
const std = @import("std");
const audio = @import("audio.zig");
const rl = @import("raylib");
const SpscQueue = @import("queue.zig").SpscQueue;

pub const Frame = u64;
pub const MAX_NOTES_PER_BLOCK = 1024; // big on purpose

pub const Note = struct {
    start: Frame,
    end: Frame,
    note: u8,
};

pub const NoteMsg = union(enum) {
    on: u8,
    off: u8,
};

pub const NoteQueue = SpscQueue(NoteMsg, 16);

pub fn beatsToFrames(beats: f32, tempo: f32, sample_rate: f32) Frame {
    return @intFromFloat((60.0 / tempo) * sample_rate * beats);
}

pub fn framesToBeats(frames: Frame, tempo: f32, sample_rate: f32) f32 {
    return @as(f32, @floatFromInt(frames)) / (sample_rate * 60.0 / tempo);
}

pub fn keyToMidi(key: rl.KeyboardKey) ?u8 {
    return switch (key) {
        // --- white keys (A–L) ---
        .a => 48, // C3
        .s => 50, // D3
        .d => 52, // E3
        .f => 53, // F3
        .g => 55, // G3
        .h => 57, // A3
        .j => 59, // B3
        .k => 60, // C4
        .l => 62, // D4
        .semicolon => 64, // E4
        .apostrophe => 65, // F4

        // --- black keys (W–O) ---
        .w => 49, // C#3
        .e => 51, // D#3
        .t => 54, // F#3
        .y => 56, // G#3
        .u => 58, // A#3
        .o => 61, // C#4
        .p => 63, // D#4
        else => null,
    };
}

pub const Player = struct {
    notes: std.ArrayListUnmanaged(Note) = .{},

    pub fn init(alloc: std.mem.Allocator, notes_in: []const Note) !Player {
        var notes: std.ArrayListUnmanaged(Note) = .{};
        try notes.appendSlice(alloc, notes_in);
        return .{ .notes = notes };
    }

    pub fn deinit(self: *Player, alloc: std.mem.Allocator) void {
        self.notes.deinit(alloc);
    }

    pub fn clear(self: *Player) void {
        self.notes.clearRetainingCapacity();
    }

    pub fn appendNotes(self: *Player, alloc: std.mem.Allocator, new_notes: []const Note) !void {
        try self.notes.appendSlice(alloc, new_notes);
    }

    pub fn advance(self: *Player, start: Frame, end: Frame, out: []NoteMsg) usize {
        std.debug.assert(end >= start);
        std.debug.assert(end - start < 8192);

        var count: usize = 0;

        for (self.notes.items) |n| {
            // TODO check what happens when advance() is called on the latter boundary wrt <, <=
            // so what if pre_accum == n.start... is this even a problem?
            //
            // TODO better array that doesn't need manual counting?
            if (start <= n.start and n.start < end) {
                if (count < out.len) {
                    std.debug.print("on: {}\n", .{n});
                    out[count] = .{ .on = n.note };
                    count += 1;
                }
            }
            if (start <= n.end and n.end < end) {
                if (count < out.len) {
                    std.debug.print("on: {}\n", .{n});
                    std.debug.print("off: {}\n", .{n});
                    out[count] = .{ .off = n.note };
                    count += 1;
                }
            }
        }
        return count;
    }
};

File: app/src/midi.rs
//! Port of `midi.zig`.

use crate::input::Key;

pub type Frame = u64;

#[derive(Clone, Copy, Debug)]
pub struct Note {
    pub start: Frame,
    pub end: Frame,
    pub note: u8,
}

pub fn beats_to_frames(beats: f32, tempo: f32, sample_rate: f32) -> Frame {
    ((60.0 / tempo) * sample_rate * beats) as Frame
}

pub fn frames_to_beats(frames: Frame, tempo: f32, sample_rate: f32) -> f32 {
    frames as f32 / (sample_rate * 60.0 / tempo)
}

/// QWERTY -> MIDI note (one octave + black keys), mirrors `midi.keyToMidi`.
pub fn key_to_midi(key: Key) -> Option<u8> {
    use Key::*;
    Some(match key {
        // white keys (A–F row)
        A => 48,          // C3
        S => 50,          // D3
        D => 52,          // E3
        F => 53,          // F3
        G => 55,          // G3
        H => 57,          // A3
        J => 59,          // B3
        K => 60,          // C4
        L => 62,          // D4
        Semicolon => 64,  // E4
        Apostrophe => 65, // F4
        // black keys (W–P row)
        W => 49, // C#3
        E => 51, // D#3
        T => 54, // F#3
        Y => 56, // G#3
        U => 58, // A#3
        O => 61, // C#4
        P => 63, // D#4
        _ => return None,
    })
}
